### ISSUE TITLE
Performance drops significantly with 2 federated code

### ISSUE THREAD

**Issue by PhungLai728** [tensorflow/federated#851]
Hi there, 
I am using federated setting for my customize dataset. I try the same data, same model (load pretrained model). However, results are very different between two code as follow:
1. 41.9% accuracy with `https://github.com/tensorflow/federated/blob/master/docs/tutorials/federated_learning_for_text_generation.ipynb`
2. 17% accuracy with `https://github.com/tensorflow/federated/blob/master/tensorflow_federated/python/research/optimization/shakespeare/run_federated.py`
I was trying different configurations for each model and the accuracy reported here is the best for each model.
I wonder whether the way of aggregating and assigning weights between clients and server are different between these code that makes the performances different.  
Have anyone got the same issues and is there any reason ? Thanks

**Comment by zcharles8** [comment]
Hi @PhungLai728. Can you describe what hyperparameters you were using with the two approaches? The choice of hyperparameters can make a huge difference, especially the choice of learning rates.

As you note, the tutorial uses a pre-trained RNN. However, currently `shakespeare/run_federated.py` trains from scratch. Did you also use the pre-trained model in `shakespeare/run_federated.py`?

As for differences in the code, the tutorial uses `tff.learning.build_federated_averaging_process` while `shakespeare/run_federated.py` uses `https://github.com/tensorflow/federated/blob/master/tensorflow_federated/python/research/optimization/shared/fed_avg_schedule.py`. While there are a few subtle differences between them, the primary difference is that the latter allows for learning rate scheduling. In the absence of using that, the two should perform comparably, barring randomness due to which clients are sampled in each round.

**Comment by PhungLai728** [comment]
Hi there, 
For both, I used pretrained model. And I mostly changed the learning rate for each model and choose the best one. The learning rate I tried is [0.01, 0.05, 0.1]. I also try learning rate schedule for the `https://github.com/tensorflow/federated/blob/master/docs/tutorials/federated_learning_for_text_generation.ipynb`. My dataset has 208 users, and I used 50 clients for each training step.

**Comment by zcharles8** [comment]
Thanks for the follow-up. It's difficult to tell what's happening without seeing the code snippets. The `shakespeare/run_federated.py` has a whole suite of flags with specific default values, so it could be that there's still some difference in configuration between the two.

That being said, `shakespeare/run_federated.py` was designed for a specific dataset and model. If you're finding success in using the tutorial's setup, I would encourage you to continue using that  to train your model.

Currently, I don't see any indication of a bug in the `shakespeare/run_federated.py`, but please let me know if you still believe there is one.

**Comment by zcharles8** [comment]
Hi @PhungLai728. I'm closing the issue for now, but if you see evidence that there is a bug in `shakespeare/run_federated.py`, please feel free to reopen this issue.
